pyccel/codegen/printing/cwrappercode.py

```diff
@@ -330,6 +333,193 @@ def optional_element_management(self, used_names, a, collect_var):
+        arg_names         = [a.name for a in funcs[0].arguments]
+        keyword_list_name = self.get_new_name(used_names,'kwlist')
+        keyword_list      = PyArgKeywords(keyword_list_name, arg_names)
+        wrapper_body              = [keyword_list]
+
+        wrapper_body_translations = []
+        body_tmp = []
+
+        # To store the mini function responsible of collecting value and calling interfaces functions and return the builded value
+        funcs_def = []
+        default_value = {} # dict to collect all initialisation needed in the wrapper
+        check_var = Variable(dtype = NativeInteger(), name = self.get_new_name(used_names , "check"))
+        wrapper_vars[check_var.name] = check_var
+        types_dict = {} #dict to collect each variable possible type and the and the corresponding flags
+        # Managing the body of wrapper
+        # TODO split or re use exisiting functiond in the wrapper
+        for func in funcs :
+            mini_wrapper_func_body = []
+            res_args = []
+            mini_wrapper_func_vars = {}
+            flags = 0
+
+            # Loop for all args in every functions and create the corresponding condition and body
+            # TODO add array management like in the wrapper of simple function
+            for a, b in zip(parse_args, func.arguments):
+                check = Type_Check(b, a) # get check type function
+                assign = [Assign(b, self.get_collect_function_call(b, a))] # get collect function
+                mini_wrapper_func_vars[b.name] = b
+
+                flag_value = flags_registry[(b.dtype, b.precision)]
+                flags = (flags << 4) + flag_value  # shift by 4 to the left
+                # Managing valued variable
+                if isinstance(b, ValuedVariable):
+                    check = PyccelAssociativeParenthesis(PyccelOr(PyccelEq(VariableAddress(a), VariableAddress(Py_None)), check))
+                    default_value[a.name] =  self.get_default_assign(a, b)
+
+                    if not b.is_optional :
+                        assign = [Assign(b, IfTernaryOperator(PyccelNe(VariableAddress(a), VariableAddress(Py_None)),
+                            self.get_collect_function_call(b, a), b.value))]
+                    else :  # Managing optional variable
+                        tmp_var = Variable(dtype=b.dtype, name = self.get_new_name(used_names, b.name+"_tmp"))
+                        mini_wrapper_func_vars[tmp_var.name] = tmp_var
+                        assign = [Assign(tmp_var, self.get_collect_function_call(tmp_var, a)), Assign(VariableAddress(b), VariableAddress(tmp_var))]
+                        assign = [If((PyccelEq(VariableAddress(a), VariableAddress(Py_None)),
+                                    [Assign(VariableAddress(b), b.value)]), (LiteralTrue(), assign))]
+
+                types_dict.setdefault(b, set()).add((b, check, flag_value)) # collect variable type for each arguments
```

Comment:
dict.setdefault(key, default=None)
key − This is the key to be searched.
default − This is the Value to be returned in case key is not found.

in the first run, key don't exist because the dict is empty  so it will  create key : value and return an empty set()
where value = set()

Reply:
ok. Is it faster if we do:
`types_dict[b] = set()`
before the loop and then just use the normal accessor? It seems more readable to me too

Reply:
it wont work 
```
dict = {}
-->func1(int  a)  
d[a]  = set()
d[a].add((a,   is_int,   flag))
--->func2(float a)
d[a] = set()
d[a].add((a,   is_float,   flag))
--------------
print(dict)
{a : (a,   is_float,   flag)}
and it should be 
{a : ((a,   is_int,   flag)  ,(a,  is_float, flag))}

```

Reply:
Ah yes, the loops are in the wrong order to be able to do this. You would have to do:
```
types_dict = {a:set() for a in func[0].arguments}
```

Reply:
thats not the same

Reply:
Meaning it doesn't get the right results?
I thought:
```

Reply:
It's just a question of syntax really. Unless there are other people who don't understand this line either then you can leave it as it is if you prefer it